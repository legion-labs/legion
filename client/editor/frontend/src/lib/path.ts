/** Dummy alias for string */
export type Path = string;

/**
 * Detects the main path separator / or \\
 * If both seperators are present in the string
 * the main separator will be assumed to be the first found
 */
export function detectMainPathSeparator(path: Path): string | null {
  for (const c of path) {
    if (c === "/") {
      return "/";
    }

    if (c === "\\") {
      return "\\";
    }
  }

  return null;
}

/** Split A path into components. Empty parts are removed. */
export function components(path: Path): string[] {
  // We assume / as the path seperator if none detected
  const pathSeparator = detectMainPathSeparator(path) || "/";

  return path.split(pathSeparator).filter(Boolean);
}

/** Extract the file name from a path */
export function fileName(path: Path): string | null {
  const parts = components(path);

  if (!parts.length) {
    return null;
  }

  return parts[parts.length - 1];
}

/** Extract the extension from a path */
export function extension(path: Path): string | null {
  const pathFileName = fileName(path);

  if (!pathFileName) {
    return null;
  }

  const pathFileNameParts = pathFileName.split(".").filter(Boolean);

  // Having only 0 or 1 part means there was no `.` found (and therefore no extensions)
  // _or_ that the file name starts with a `.`
  if (pathFileNameParts.length <= 1) {
    return null;
  }

  return pathFileNameParts.reverse()[0];
}
