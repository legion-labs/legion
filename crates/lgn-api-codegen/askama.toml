[general]
# Directories to search for templates, relative to the crate root.
dirs = [
    "src/rust/templates",
    "src/typescript/templates",
    "src/python/templates",
]
whitespace = "suppress"
