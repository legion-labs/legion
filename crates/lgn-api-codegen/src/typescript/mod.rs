use std::{
    borrow::Cow,
    fs::{read_to_string, File},
    io::Write,
    path::Path,
    process::Command,
};

use anyhow::anyhow;
use askama::Template;
use tempfile::tempdir;
use which::which;

use crate::{
    api_types::{GenerationContext, MediaType, Type},
    errors::Error,
    Language, Result, TypeScriptOptions,
};

mod filters;

#[derive(askama::Template)]
#[template(path = "index.ts.jinja", escape = "none")]
struct TypeScriptIndexTemplate<'a> {
    pub ctx: &'a TypeScriptGenerationContext,
}

#[derive(askama::Template)]
#[template(path = "package.json.jinja", escape = "none")]
struct TypeScriptPackageTemplate;

pub type TypeScriptGenerationContext = GenerationContext<TypeScriptOptions>;

fn format_typescript<'a, 'b>(
    options: &TypeScriptOptions,
    content: &'a str,
    temp_file_name: &'b str,
) -> Result<Cow<'a, str>> {
    let prettier_config_path = options
        .prettier_config_path
        .clone()
        .map(|path| path.to_string_lossy().into_owned());

    let dir =
        tempdir().map_err(|_err| Error::TypeScriptFormat(anyhow!("couldn't create a tmp dir")))?;

    let file_path = dir.path().join(temp_file_name);

    let mut file = File::create(&file_path)?;

    write!(file, "{}", content)
        .map_err(|_err| Error::TypeScriptFormat(anyhow!("couldn't write content to tmp file")))?;

    let binary_path = match which("npx") {
        Err(_) => return Ok(content.into()),
        Ok(path) => path,
    };

    let mut command = Command::new(binary_path);

    let tmp_file_path = file_path.as_path().to_string_lossy();

    let mut args = vec![
        "--yes",
        "prettier",
        "--loglevel",
        "silent",
        "--write",
        &tmp_file_path,
    ];

    match prettier_config_path {
        None => args.push("--no-config"),
        Some(ref path) => {
            args.push("--config");
            args.push(path);
        }
    };

    command.args(&args);

    let status = command.status().map_err(|error| {
        Error::TypeScriptFormat(anyhow!(
            "npx prettier command was not successful: {}",
            error
        ))
    })?;

    if status.success() {
        let formatted_content = read_to_string(&file_path)
            .map_err(|_err| Error::TypeScriptFormat(anyhow!("couldn't read tmp file")))?;

        Ok(formatted_content.into())
    } else {
        Err(Error::TypeScriptFormat(anyhow!(
            "npx prettier command was not successful: {}",
            status
        )))
    }
}

fn generate_index_content(ctx: &TypeScriptGenerationContext) -> Result<String> {
    let mut content = TypeScriptIndexTemplate { ctx }.render()?;

    if !ctx.options.skip_format {
        content = format_typescript(&ctx.options, &content, "index.ts")?.into_owned();
    }

    Ok(content)
}

fn generate_package_content(options: &TypeScriptOptions) -> Result<String> {
    let mut content = TypeScriptPackageTemplate.render()?;

    if !options.skip_format {
        content = format_typescript(options, &content, "package.json")?.into_owned();
    }

    Ok(content)
}

impl Language {
    pub(crate) fn generate_typescript(
        ctx: GenerationContext,
        options: TypeScriptOptions,
        output_dir: &Path,
    ) -> Result<()> {
        std::fs::create_dir_all(output_dir)?;

        let ctx = ctx.with_options(options);
        let output_file = output_dir.join("index.ts");
        let content = generate_index_content(&ctx)?;

        std::fs::write(output_file, content)?;

        if ctx.options.with_package_json {
            let output_file = output_dir.join("package.json");
            let content = generate_package_content(&ctx.options)?;

            std::fs::write(output_file, content)?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use crate::{openapi_loader::OpenApiRefLocation, visitor::Visitor, OpenApiLoader};

    use super::*;

    #[test]
    fn test_ts_index_generation() {
        let loader = OpenApiLoader::default();
        let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../tests/api-codegen")
            .canonicalize()
            .unwrap();
        let openapi = loader
            .load_openapi(OpenApiRefLocation::new(&root, "cars.yaml".into()))
            .unwrap();
        let ctx = Visitor::new(root)
            .visit(&[openapi.clone()])
            .unwrap()
            .with_options(TypeScriptOptions::default());
        let content = generate_index_content(&ctx).unwrap();

        insta::assert_snapshot!(content);
    }

    #[test]
    fn test_ts_package_generation() {
        let content = generate_package_content(&TypeScriptOptions::default()).unwrap();

        insta::assert_snapshot!(content);
    }
}
