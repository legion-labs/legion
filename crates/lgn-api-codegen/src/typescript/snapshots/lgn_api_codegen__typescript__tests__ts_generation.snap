---
source: crates/lgn-api-codegen/src/typescript/mod.rs
expression: content
---
// Auto-generated file for Api: Test API : 1.0.0
//
// This is a sample OpenAPI 3.0 specification.
//

import qs from "qs";

// ---------- Api ----------

export interface Api {

    
        
        getCars(input: {
    params: {
        "space-id": string;
        },
    
    query: {
        "names"?: string[];
        "q"?: string;
        },
    
    
    
    }): Promise<GetCarsResponse>;
    
        
        createCar(input: {
    params: {
        "space-id": string;
        },
    
    
    headers: {
        "span-id"?: string;
        },
    
    
    body: CarModel,}): Promise<CreateCarResponse>;
    

    
        
        getCar(input: {
    params: {
        "space-id": string;
        "car-id": bigint;
        },
    
    
    
    }): Promise<GetCarResponse>;
    
        
        deleteCar(input: {
    params: {
        "space-id": string;
        "car-id": bigint;
        },
    
    
    
    }): Promise<DeleteCarResponse>;
    

    
        
        testBinary(input: {
    params: {
        "space-id": string;
        },
    
    
    
    body: Blob,}): Promise<TestBinaryResponse>;
    

    
        
        testOneOf(input: {
    params: {
        },
    
    
    
    }): Promise<TestOneOfResponse>;
    

}

// ---------- Errors ----------

declare global {
  interface ErrorConstructor {
    captureStackTrace(thisArg: unknown, func: unknown): void;
  }
}

export class InternalError extends Error {
  constructor(message: string) {
    super(`Internal error: ${message}`);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, InternalError)
    }

    this.name = "InternalError";
  }
}

// ---------- Client ----------

type Fetch = typeof globalThis.fetch;

const defaultFetch = globalThis.fetch;

export class Client implements Api {
    #baseUri: string | URL;
    #fetch: Fetch;

    constructor({ baseUri, fetch = defaultFetch }: { baseUri: string | URL, fetch?: Fetch }) {
        this.#baseUri = baseUri;
        this.#fetch = fetch;
    }

    
        
            

            async getCars(input: {
    params: {
        "space-id": string;
        },
    
    query: {
        "names"?: string[];
        "q"?: string;
        },
    
    
    
    }): Promise<GetCarsResponse> {
                const url = new URL(`${this.#baseUri}/v1/spaces/${input.params["space-id"]}/car-service/cars`);

                const query: GetCarsQuery = { 
                        "names": input.query["names"],
                        "q": input.query["q"],
                        };

                    url.search = `?${qs.stringify(query)}`;
                

                const headers = new Headers();

                


                

                const response = await this.#fetch(url.toString(), {
                    method: "GET",
                    headers,
                    
                });

                return GetCarsResponse.fromResponse(response);
            }
        
            

            async createCar(input: {
    params: {
        "space-id": string;
        },
    
    
    headers: {
        "span-id"?: string;
        },
    
    
    body: CarModel,}): Promise<CreateCarResponse> {
                const url = new URL(`${this.#baseUri}/v1/spaces/${input.params["space-id"]}/car-service/cars`);

                

                const headers = new Headers();

                
                    headers.set(
                        "Content-Type",
                        "application/json",
                    );
                


                

                const response = await this.#fetch(url.toString(), {
                    method: "POST",
                    headers,
                    
                        
                                body: JSON.stringify(input.body),
                            
                    
                    
                });

                return CreateCarResponse.fromResponse(response);
            }
        
    
        
            

            async getCar(input: {
    params: {
        "space-id": string;
        "car-id": bigint;
        },
    
    
    
    }): Promise<GetCarResponse> {
                const url = new URL(`${this.#baseUri}/v1/spaces/${input.params["space-id"]}/car-service/cars/${input.params["car-id"]}`);

                

                const headers = new Headers();

                


                

                const response = await this.#fetch(url.toString(), {
                    method: "GET",
                    headers,
                    
                });

                return GetCarResponse.fromResponse(response);
            }
        
            

            async deleteCar(input: {
    params: {
        "space-id": string;
        "car-id": bigint;
        },
    
    
    
    }): Promise<DeleteCarResponse> {
                const url = new URL(`${this.#baseUri}/v1/spaces/${input.params["space-id"]}/car-service/cars/${input.params["car-id"]}`);

                

                const headers = new Headers();

                


                

                const response = await this.#fetch(url.toString(), {
                    method: "DELETE",
                    headers,
                    
                });

                return DeleteCarResponse.fromResponse(response);
            }
        
    
        
            

            async testBinary(input: {
    params: {
        "space-id": string;
        },
    
    
    
    body: Blob,}): Promise<TestBinaryResponse> {
                const url = new URL(`${this.#baseUri}/v1/spaces/${input.params["space-id"]}/car-service/test-binary`);

                

                const headers = new Headers();

                
                    headers.set(
                        "Content-Type",
                        "application/octet-stream",
                    );
                


                

                const response = await this.#fetch(url.toString(), {
                    method: "POST",
                    headers,
                    
                        
                                body: input.body,
                        
                    
                    
                });

                return TestBinaryResponse.fromResponse(response);
            }
        
    
        
            

            async testOneOf(input: {
    params: {
        },
    
    
    
    }): Promise<TestOneOfResponse> {
                const url = new URL(`${this.#baseUri}/test-one-of`);

                

                const headers = new Headers();

                


                

                const response = await this.#fetch(url.toString(), {
                    method: "GET",
                    headers,
                    
                });

                return TestOneOfResponse.fromResponse(response);
            }
        
    
}

// ---------- Models ----------


    
            /** The car color. */
            export type CarColorModel = (| "red"| "blue"| "yellow");
        

    
            
            export type CarModel = {
                
                    
                    "id": bigint;
                    
                    "name": string;
                    /** The car color. */
                    "color": CarColorModel;
                    
                    "is_new": boolean;
                    
                    "extra"?: Blob;
            };
    

    
            
            export type PetModel = {
                
                    
                    "name"?: string;
            };
    

    
            
            export type TestOneOfResponseModel = (| { type: "option1", value: PetModel }| { type: "option2", value: CarModel });
        


// ---------- Parameters ----------


        
        export type GetCarsQuery = {
                "names"?: string[];
                "q"?: string;
                
            };
        
        
        
    

        
        
        
        
    

        
        
    

        
        
    


// ---------- Responses ----------


    
        
        
        export type GetCarsResponse = (| { /** List of cars. */ type: "200", value: CarModel[] });

        export const GetCarsResponse = {
            async fromResponse(response: Response): Promise<GetCarsResponse> {
                
                    switch (response.status) {
                        
                            case 200: { 
                                return {type: "200", value:
                                                await response.json()
                                            }
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    
        
        
        export type CreateCarResponse = (| { /** Created. */ type: "201" });

        export const CreateCarResponse = {
            async fromResponse(response: Response): Promise<CreateCarResponse> {
                
                    switch (response.status) {
                        
                            case 201: { 
                                return {type: "201", value:}
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    

    
        
        
        export type GetCarResponse = (| { /** A car. */ type: "200", value: CarModel }| { /** Car not found. */ type: "404" });

        export const GetCarResponse = {
            async fromResponse(response: Response): Promise<GetCarResponse> {
                
                    switch (response.status) {
                        
                            case 200: { 
                                return {type: "200", value:
                                                await response.json()
                                            }
                            }
                            case 404: { 
                                return {type: "404", value:}
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    
        
        
        export type DeleteCarResponse = (| { /** Car deleted. */ type: "200" }| { /** Car not found. */ type: "404" });

        export const DeleteCarResponse = {
            async fromResponse(response: Response): Promise<DeleteCarResponse> {
                
                    switch (response.status) {
                        
                            case 200: { 
                                return {type: "200", value:}
                            }
                            case 404: { 
                                return {type: "404", value:}
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    

    
        
        
        export type TestBinaryResponse = (| { /** Ok. */ type: "200", value: Blob });

        export const TestBinaryResponse = {
            async fromResponse(response: Response): Promise<TestBinaryResponse> {
                
                    switch (response.status) {
                        
                            case 200: { 
                                return {type: "200", value:
                                                await response.blob()
                                        }
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    

    
        
        
        export type TestOneOfResponse = (| { /** Ok. */ type: "200", value: TestOneOfResponseModel });

        export const TestOneOfResponse = {
            async fromResponse(response: Response): Promise<TestOneOfResponse> {
                
                    switch (response.status) {
                        
                            case 200: { 
                                return {type: "200", value:
                                                await response.json()
                                            }
                            }default: {
                            throw new InternalError(`unexpected status code: ${status}`);
                        }
                    }
                
            } 
        }
    

